tests/h/views/api_groups_test.py

```diff
@@ -12,44 +12,65 @@
 from h.services.group import GroupService
 
 
+@pytest.mark.usefixtures('GroupsJSONPresenter')
 class TestGroups(object):
-    def test_groups_proxies_to_service(self, pyramid_request, list_groups_service):
-        views.groups(pyramid_request)
 
-        list_groups_service.all_groups.assert_called_once()
+    def test_all_groups_proxies_to_service(self, anonymous_request, list_groups_service):
+        views.groups(anonymous_request)
 
-    def test_groups_passes_authority_parameter(self, pyramid_request, list_groups_service):
-        pyramid_request.params = {'authority': 'foo.com'}
+        list_groups_service.all_groups.assert_called_once_with(
+            user=None,
+            authority=None,
+            document_uri=None
+        )
 
-        views.groups(pyramid_request)
+    def test_all_groups_proxies_authority_param(self, anonymous_request, list_groups_service):
+        anonymous_request.params['authority'] = 'foo.com'
+        views.groups(anonymous_request)
 
-        c_args, c_kwargs = list_groups_service.all_groups.call_args
-        assert c_kwargs['authority'] == 'foo.com'
+        list_groups_service.all_groups.assert_called_once_with(
+            user=None,
+            authority='foo.com',
+            document_uri=None
+        )
 
-    def test_groups_passes_document_uri_parameter(self, pyramid_request, list_groups_service):
-        pyramid_request.params = {'document_uri': 'foo.example.com'}
+    def test_all_groups_proxies_document_uri_param(self, anonymous_request, list_groups_service):
+        anonymous_request.params['document_uri'] = 'http://example.com/thisthing.html'
+        views.groups(anonymous_request)
 
-        views.groups(pyramid_request)
+        list_groups_service.all_groups.assert_called_once_with(
+            user=None,
+            authority=None,
+            document_uri='http://example.com/thisthing.html'
+        )
 
-        c_args, c_kwargs = list_groups_service.all_groups.call_args
-        assert c_kwargs['document_uri'] == 'foo.example.com'
+    def test_uses_presenter_for_formatting(self, anonymous_request, open_groups, list_groups_service, GroupsJSONPresenter):  # noqa: N803
+        list_groups_service.all_groups.return_value = open_groups
+
+        views.groups(anonymous_request)
+
+        assert GroupsJSONPresenter.call_count == 1
 
     @pytest.fixture
-    def pyramid_request(self, pyramid_request, user):
-        pyramid_request.user = user
-        pyramid_request.json_body = {}
+    def pyramid_request(self, pyramid_request):
+        pyramid_request.route_url = mock.Mock(return_value='/groups/foo')
```

Comment:
Ideally we should use `routes` fixtures instead of mocking `route_url`.

Reply:
Hi! Can you elaborate a bit here? I think I see how I could extend the `routes` fixture (which I finally located in `tests/h/views/main_test`...I think?) but I'm not sure how to apply that to `route_url` the method.

Reply:
Ah no, there's no one routes fixture, rather routes fixtures are a pattern used throughout the tests. You'll find lots of fixtures called `routes` defined in different test classes and modules and used within those class / modules. So you could define your own routes fixture in this class. Something like:

```
    @pytest.fixture
    def routes(self, pyramid_config):
        pyramid_config.add_route('group_read', '/groups/{pubid}/{slug}')
```

It's a fixture that modifies the `pyramid_config` fixture that in turn modifies the `pyramid_request` fixture! So the end result is that the dummy `pyramid_request` receives the route definition.

Reply:
But does `pyramid_request` have a `route_url` method? I'm unclear of how to mock that such that it interfaces with the `routes` fixture.

Reply:
Wait a sec, `api_groups.py` doesn't call `route_url`! It just passes it in to the (mock) `GroupJSONPresenter`. So this isn't needed at all - the `pyramid_request` fixture function here can just be deleted.

For the record though `pyramid_request` _does_ have a `route_url()` method. It's not a mock object, it's a [DummyRequest](https://docs.pylonsproject.org/projects/pyramid//en/latest/api/testing.html?highlight=dummyrequest#pyramid.testing.DummyRequest). Our `pyramid_request` fixture just wraps Pyramid's `DummyRequest` class making it available as a pytest fixture:

https://github.com/hypothesis/h/blob/74c60e9f7c15222731cc583af7728bd039a56dd0/tests/h/conftest.py#L227-L239

If you create a `routes` fixture that uses the `pyramid_config` fixture and calls `pyramid_config.add_route()` that route that you added ends up on the dummy `pyramid_request` object because our `pyramid_config` fixture applies itself to the `pyramid_request` fixture:

https://github.com/hypothesis/h/blob/74c60e9f7c15222731cc583af7728bd039a56dd0/tests/h/conftest.py#L215-L224

(Yes, fixtures can mess with other fixtures.)

So if `api_groups.py` were actually calling `route_url()` this would work:

```diff
diff --git a/tests/h/views/api_groups_test.py b/tests/h/views/api_groups_test.py
index 66374e62c..66d8c0d04 100644
--- a/tests/h/views/api_groups_test.py
+++ b/tests/h/views/api_groups_test.py
@@ -14,6 +14,7 @@ from h.services.group import GroupService
 pytestmark = pytest.mark.usefixtures('GroupsJSONPresenter')
 
 
+@pytest.mark.usefixtures('routes')
 class TestGroups(object):
 
     def test_all_groups_proxies_to_service(self, anonymous_request, list_groups_service):
@@ -60,9 +61,8 @@ class TestGroups(object):
         assert result == GroupsJSONPresenter(open_groups, anonymous_request).asdicts.return_value
 
     @pytest.fixture
-    def pyramid_request(self, pyramid_request):
-        pyramid_request.route_url = mock.Mock(return_value='/groups/foo')
-        return pyramid_request
+    def routes(self, pyramid_config):
+        pyramid_config.add_route('group_read', '/groups/{pubid}')
 
     @pytest.fixture
     def open_groups(self, factories):
```